File: clients/android/core-model/src/main/kotlin/de/juloc/anilingo/core/model/OfflineLibraryModels.kt
package de.juloc.anilingo.core.model

/**
 * Client mirror of the server's offline Book/Novel library contract
 * (`/api/client/v1/offline-library` endpoints, capability `offlineLibrary`,
 * see `docs/OFFLINE_LIBRARY.md`). Books and Novels share one server model; these
 * types are generic over "work -> volume -> chapter" and carry no anime/media
 * fields, mirroring `Features/OfflineLibrary` on the server exactly.
 */
data class ClientOfflineLibraryVolume(
    val volumeId: String,
    val number: Int,
    val title: String?,
    val kind: String,
    val coverAssetUrl: String?,
)

/** One chapter reference inside a manifest; [hash] is the differential-download key. */
data class ClientOfflineLibraryChapterRef(
    val chapterId: String,
    val volumeId: String,
    val number: Int,
    val title: String,
    val hash: String,
    val hasContent: Boolean,
    val hasTranslation: Boolean,
)

data class ClientOfflineLibraryManifest(
    val workId: String,
    val schemaVersion: Int,
    val contentVersion: String,
    val title: String,
    val author: String?,
    val description: String?,
    val coverAssetUrl: String?,
    val issuedAtUtc: String,
    val volumes: List<ClientOfflineLibraryVolume>,
    val chapters: List<ClientOfflineLibraryChapterRef>,
)

/** Manifest plus the exact JSON it was parsed from, so it can be stored and re-read offline. */
data class OfflineLibraryManifestPackage(
    val manifest: ClientOfflineLibraryManifest,
    val json: String,
)

data class ClientOfflineLibraryInlineRun(
    val text: String,
    val ruby: String?,
    val emphasis: Boolean,
    val strong: Boolean,
)

data class ClientOfflineLibraryContentBlock(
    val kind: String,
    val level: Int,
    val runs: List<ClientOfflineLibraryInlineRun>,
    val imageAssetUrl: String?,
    val imageAlt: String?,
)

data class ClientOfflineLibraryTranslation(
    val targetLanguage: String,
    val text: String,
)

/**
 * [hash] is repeated here (equal to the manifest's ref for this chapter) so a
 * client can detect a version race: the chapter changed again after the
 * manifest was fetched but before this payload landed.
 */
data class ClientOfflineLibraryChapterPayload(
    val chapterId: String,
    val workId: String,
    val volumeId: String,
    val number: Int,
    val title: String,
    val hash: String,
    val originalText: String,
    val blocks: List<ClientOfflineLibraryContentBlock>,
    val translations: List<ClientOfflineLibraryTranslation>,
)

/** Payload plus the exact JSON it was parsed from, so it can be stored and re-read offline. */
data class OfflineLibraryChapterPackage(
    val payload: ClientOfflineLibraryChapterPayload,
    val json: String,
)

/** One offline reading-position checkpoint for `POST /offline-library/sync`. */
data class OfflineLibraryProgressEvent(
    val clientEventId: String,
    val workId: String,
    val chapterId: String,
    val positionPermille: Int,
    val anchorLanguage: String?,
    val anchorParagraphIndex: Int?,
    val anchorOffset: Int,
    val clientTimestampUtc: String,
)

/** [type] is `"upsert"` (add/edit) or `"remove"`; the bookmark id is client-supplied. */
data class OfflineLibraryBookmarkEvent(
    val clientEventId: String,
    val bookmarkId: String,
    val type: String,
    val workId: String,
    val chapterId: String,
    val language: String?,
    val positionPermille: Int,
    val paragraphIndex: Int?,
    val characterOffset: Int,
    val anchorText: String?,
    val label: String?,
    val style: String?,
    val color: String?,
    val clientTimestampUtc: String,
)

data class OfflineLibraryProgressResult(
    val clientEventId: String,
    val workId: String,
    val outcome: String,
    val positionPermille: Int?,
)

data class OfflineLibraryBookmarkResult(
    val clientEventId: String,
    val bookmarkId: String,
    val outcome: String,
)

data class OfflineLibrarySyncResult(
    val progress: List<OfflineLibraryProgressResult>,
    val bookmarks: List<OfflineLibraryBookmarkResult>,
)

File: src/AniLingo.Web/Features/OfflineLibrary/OfflineLibraryModels.cs
namespace AniLingo.Web.Features.OfflineLibrary;

/// <summary>
/// Tombstone of a bookmark removed through offline sync. Kept so a stale
/// "add"/"edit" replay arriving after a newer removal can never resurrect the
/// bookmark (last-writer-wins by client timestamp, mirroring
/// <c>OfflineProgressReconciler</c>'s monotonic rule for playback progress).
/// One row per bookmark id; re-adding the same id later replaces it.
/// </summary>
public sealed class NovelBookmarkTombstone
{
    /// <summary>Same id the live bookmark had; a primary key, not a foreign key (the bookmark is gone).</summary>
    public Guid BookmarkId { get; set; }
    public string ProfileId { get; set; } = "";
    public Guid WorkId { get; set; }

    /// <summary>Client-asserted last-writer-wins timestamp of the removal.</summary>
    public DateTime DeletedAtUtc { get; set; }

    /// <summary>When the server first recorded this tombstone (for future pruning).</summary>
    public DateTime RecordedAtUtc { get; set; } = DateTime.UtcNow;
}

/// <summary>One volume of a work as it appears in the offline manifest.</summary>
public sealed record OfflineLibraryVolume(
    Guid VolumeId,
    int Number,
    string? Title,
    string Kind,
    string? CoverAssetUrl);

/// <summary>One chapter's identity, order and differential version.</summary>
public sealed record OfflineLibraryChapterRef(
    Guid ChapterId,
    Guid VolumeId,
    int Number,
    string Title,
    string Hash,
    bool HasContent,
    bool HasTranslation);

/// <summary>
/// Canonical versioned offline package manifest for one work: enough for a
/// client to plan a differential download (per-chapter hash comparison) and
/// render a library/detail "offline available" state without contacting the
/// server again until a chapter hash changes.
/// </summary>
public sealed record OfflineLibraryManifest(
    Guid WorkId,
    int SchemaVersion,
    string ContentVersion,
    string Title,
    string? Author,
    string? Description,
    string? CoverAssetUrl,
    DateTime IssuedAtUtc,
    IReadOnlyList<OfflineLibraryVolume> Volumes,
    IReadOnlyList<OfflineLibraryChapterRef> Chapters);

/// <summary>One inline text/heading/image block, wire-shaped for offline storage (mirrors <c>NovelContentBlock</c>).</summary>
public sealed record OfflineLibraryContentBlock(
    string Kind,
    int Level,
    IReadOnlyList<OfflineLibraryInlineRun>? Runs,
    string? ImageAssetUrl,
    string? ImageAlt);

public sealed record OfflineLibraryInlineRun(
    string Text,
    string? Ruby,
    bool Emphasis,
    bool Strong);

/// <summary>One cached translation of a chapter.</summary>
public sealed record OfflineLibraryTranslation(
    string TargetLanguage,
    string Text);

/// <summary>
/// Full content of one chapter for offline storage: original text/blocks plus
/// every cached translation, so the client never needs to pick a language up
/// front. Referenced images are content-addressed asset URLs the client fetches
/// and stores itself (never embedded as bytes here).
/// </summary>
public sealed record OfflineLibraryChapterPayload(
    Guid ChapterId,
    Guid WorkId,
    Guid VolumeId,
    int Number,
    string Title,
    string Hash,
    string OriginalText,
    IReadOnlyList<OfflineLibraryContentBlock> Blocks,
    IReadOnlyList<OfflineLibraryTranslation> Translations);

/// <summary>View model for the reusable "Offline speichern" partial (Pages/Shared/_OfflineLibraryAction.cshtml).</summary>
public sealed record OfflineLibraryActionViewModel(Guid WorkId, string Title);
